cunumeric/module.py

```diff
@@ -1257,6 +1278,44 @@ def check_shape_dtype(
     return converted, ArrayInfo(ndim, shape, dtype)
 
 
+def _block(arr, cur_depth, depth):
+    if len(arr) > 0 and cur_depth < depth:
+        groups = []
+        cur_idx = None
+        cur_siblings = []
+        # collect subtrees by comparing indicies[0:cur_depth+1]
+        for child in arr:
+            if cur_idx != child[0][: cur_depth + 1]:
+                if len(cur_siblings) > 0:
+                    groups.append(cur_siblings)
+                    cur_siblings = []
+                cur_idx = child[0][: cur_depth + 1]
+            cur_siblings.append(child)
+        groups.append(cur_siblings)
+        # do recursive calls for each subtree
+        # and collect the results of subtrees as siblings
+        siblings = list(_block(each, cur_depth + 1, depth) for each in groups)
+        inputs = siblings
+    elif cur_depth == depth:
+        inputs = list(
+            ndarray.convert_to_cunumeric_ndarray(inp) for inp in arr[0][1]
+        )
+
+    # Computes the maximum number of dimensions for the concatenation
+    max_list = list(inp.ndim for inp in inputs)
+    max_list.append(1 + (depth - cur_depth))
+    max_ndim = _builtin_max(max_list)
+    # Append leading 1's to make elements to have the same 'ndim'
+    reshaped = list(
+        inp.reshape((1,) * (max_ndim - inp.ndim) + inp.shape)
+        if max_ndim > inp.ndim
+        else inp
+        for inp in inputs
+    )
+
+    return concatenate(reshaped, axis=-1 + (cur_depth - depth))
```

Comment:
Note that the canonical NumPy doesn't use a concatenation based implementation except for cases where the output is fairly small: https://github.com/numpy/numpy/blob/v1.22.0/numpy/core/shape_base.py#L846-L849
Given each concatenation produces a fresh array, I think we should also switch to a slicing based implementation.

Reply:
All right. I haven't thought of the method. I'll take a look and think of how we can adopt it.